service/worker/scanner/shardscanner/activities.go

```diff
@@ -0,0 +1,336 @@
+		}
+	}
+	return heartbeatDetails.Reports, nil
+}
+
+func fixShard(
+	activityCtx context.Context,
+	params FixShardActivityParams,
+	shardID int,
+	corruptedKeys store.Keys,
+	heartbeatDetails FixShardHeartbeatDetails,
+) (*FixReport, error) {
+	ctx := activityCtx.Value(params.ContextKey).(FixerContext)
+	resources := ctx.Resource
+	scope := ctx.Scope.Tagged(metrics.ActivityTypeTag(ActivityFixShard))
+	sw := scope.StartTimer(metrics.CadenceLatency)
+	defer sw.Stop()
+
+	if ctx.Hooks == nil {
+		return nil, errors.New("fixer hooks are not provided")
+	}
+
+	execManager, err := resources.GetExecutionManager(shardID)
+	if err != nil {
+		scope.IncCounter(metrics.CadenceFailures)
+		return nil, err
+	}
+
+	pr := persistence.NewPersistenceRetryer(execManager, resources.GetHistoryManager(), c.CreatePersistenceRetryPolicy())
+
+	fixer := NewFixer(
+		shardID,
+		ctx.Hooks.InvariantManager(pr, params, *ctx.Config),
+		ctx.Hooks.Iterator(resources.GetBlobstoreClient(), corruptedKeys, params, *ctx.Config),
+		resources.GetBlobstoreClient(),
+		params.ResolvedFixerWorkflowConfig.BlobstoreFlushThreshold,
+		func() { activity.RecordHeartbeat(activityCtx, heartbeatDetails) },
+	)
+	report := fixer.Fix(activityCtx)
+	if report.Result.ControlFlowFailure != nil {
+		scope.IncCounter(metrics.CadenceFailures)
+	}
+	return &report, nil
+}
+
+// ScannerEmitMetricsActivity will emit metrics for a complete run of ShardScanner
+func ScannerEmitMetricsActivity(
+	activityCtx context.Context,
+	params ScannerEmitMetricsActivityParams,
+) error {
+	scope := activityCtx.Value(params.ContextKey).(Context).Scope.Tagged(
+		metrics.ActivityTypeTag(ActivityScannerEmitMetrics),
+	)
+	scope.UpdateGauge(metrics.CadenceShardSuccessGauge, float64(params.ShardSuccessCount))
+	scope.UpdateGauge(metrics.CadenceShardFailureGauge, float64(params.ShardControlFlowFailureCount))
+
+	agg := params.AggregateReportResult
+	scope.UpdateGauge(metrics.ScannerExecutionsGauge, float64(agg.EntitiesCount))
+	scope.UpdateGauge(metrics.ScannerCorruptedGauge, float64(agg.CorruptedCount))
+	scope.UpdateGauge(metrics.ScannerCheckFailedGauge, float64(agg.CheckFailedCount))
+	for k, v := range agg.CorruptionByType {
+		scope.Tagged(metrics.InvariantTypeTag(string(k))).UpdateGauge(metrics.ScannerCorruptionByTypeGauge, float64(v))
+	}
+	shardStats := params.ShardDistributionStats
+	scope.UpdateGauge(metrics.ScannerShardSizeMaxGauge, float64(shardStats.Max))
+	scope.UpdateGauge(metrics.ScannerShardSizeMedianGauge, float64(shardStats.Median))
+	scope.UpdateGauge(metrics.ScannerShardSizeMinGauge, float64(shardStats.Min))
+	scope.UpdateGauge(metrics.ScannerShardSizeNinetyGauge, float64(shardStats.P90))
+	scope.UpdateGauge(metrics.ScannerShardSizeSeventyFiveGauge, float64(shardStats.P75))
+	scope.UpdateGauge(metrics.ScannerShardSizeTwentyFiveGauge, float64(shardStats.P25))
+	scope.UpdateGauge(metrics.ScannerShardSizeTenGauge, float64(shardStats.P10))
```

Comment:
As per my earlier comment about the scan type specific report I think right here we should have something that is basically like 

if hooks != nil {
    hooks.EmitMetrics(report)
}